src/main/java/org/jenkinsci/plugins/gitclient/GitCredentialBindings.java

```diff
@@ -2,21 +2,75 @@
 
-    String GIT_TOOL_NAME = FilenameUtils.removeExtension(GitTool.getDefaultInstallation().getGitExe());
+    void setKeyBindings(@NonNull StandardCredentials credentials);
 
-    void setKeyBindings(@Nonnull StandardCredentials credentials);
-
-    void setRunEnviornmentVariables(@Nonnull FilePath filePath, @Nonnull TaskListener listener) throws IOException, InterruptedException;
+    void setRunEnviornmentVariables(@NonNull FilePath filePath, @NonNull TaskListener listener) throws IOException, InterruptedException;
 
     GitClient getGitClientInstance(TaskListener listener) throws IOException, InterruptedException;
 
+    default String gitToolName(Run<?, ?> run, TaskListener listener) throws IOException, InterruptedException {
+
+        Executor buildExecutor = run.getExecutor();
+        ToolDescriptor<? extends GitTool> toolType;
+        if (buildExecutor != null) {
+            Node currentNode = buildExecutor.getOwner().getNode();
+            //Check node is not null
+            if (currentNode != null) {
+                if (currentNode.getNodeName().equals("")) {
+                    //It's on the master node
+                    toolType = Jenkins.get().getDescriptorByType(GitTool.DescriptorImpl.class);
+                    if (toolType != null) {
+                        ToolInstallation[] toolInstalled = toolType.getInstallations();
+                        for (ToolInstallation t : toolInstalled) {
+                            if (t.getClass().equals(GitTool.class)) {
+                                return ((GitTool) t).getGitExe();
+                            }
+                        }
+                    }
+                } else {
+                    //It's on an agent type
+                    ToolLocationNodeProperty nodeToolLocation = currentNode.getNodeProperty(ToolLocationNodeProperty.class);
+                    //Tool location property is configured
+                    if (nodeToolLocation != null) {
+                        //Get Tool locations(could be many)
+                        List<ToolLocationNodeProperty.ToolLocation> toolList = nodeToolLocation.getLocations();
+                        //Iteration
+                        for (ToolLocationNodeProperty.ToolLocation tool : toolList) {
+                            //If 'Default' Git Tool is chosen on an agent then all the Git Tools installations added
+                            // to the master nodes will be available on the agent as well
+                            //Check not an instance of jgit or jgitapache using tool descriptor not type key
+                            if (tool.getType().clazz.equals(GitTool.class)) {
+                                //Check for properties
+                                GitTool gitTool = new GitTool(tool.getName(), tool.getHome(), null);
+                                //node specific git home for git tool
+                                gitTool = gitTool.forNode(currentNode, listener);
```

Comment:
Does this work as an alternative? https://javadoc.jenkins.io/hudson/tools/ToolLocationNodeProperty.html#getHome-hudson.tools.ToolInstallation-

Reply:
[translateFor](https://javadoc.jenkins.io/hudson/tools/ToolInstallation.html#translateFor-hudson.model.Node-hudson.model.TaskListener-). It just overrides the node specific values for the git-tool instance variables,  `githome` value for a default git is `git` or `git.exe` but it could be different when the build running on an agent.